internal/cli/watch_opts.go

```diff
@@ -40,21 +44,44 @@ func (opts *WatchOpts) Watch(f Watcher) error {
 		return errors.New("no watcher provided")
 	}
 	opts.start()
+	opts.n = 2
 	for {
 		done, err := f()
-		if err != nil || done {
+		if err != nil {
+			if opts.exponentialBackoff(err) {
+				continue
+			}
+			opts.stop()
+			return err
+		}
+		if done {
 			opts.stop()
 			return err
 		}
 		if !opts.IsTerminal() {
 			if _, err = fmt.Fprint(opts.ConfigWriter(), "."); err != nil {
+				if opts.exponentialBackoff(err) {
+					continue
+				}
 				return err
 			}
 		}
 		time.Sleep(defaultWait)
 	}
 }
 
+func (opts *WatchOpts) exponentialBackoff(err error) bool {
+	// 404 and 401 are the most common errors thrown when upgrading cluster, as it is temporarily deleted
+	if opts.n <= 8 && (strings.Contains(err.Error(), "404")) || (strings.Contains(err.Error(), "401")) {
```

Comment:
string comparison for errors is a bad practice, please use go1.13 error handling, https://go.dev/blog/go1.13-errors
there are alot of examples on how to check for atlas api errors in the code base

Reply:
Thanks for the information, indeed, it wasn't the best way of checking. I improved it.